runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -269,31 +312,43 @@ public final boolean awaitTermination(int time, TimeUnit unit) throws Interrupte
 
   @Override
   public final Instant startTime() {
-    return new Instant(startTimeMs.get());
+    return new Instant(debugMetrics.getStartTimeMs());
   }
 
   @Override
   public String backendWorkerToken() {
     return backendWorkerToken;
   }
 
+  @SuppressWarnings("GuardedBy")
   @Override
-  public void shutdown() {
-    if (isShutdown.compareAndSet(false, true)) {
-      requestObserver()
-          .onError(new WindmillStreamShutdownException("Explicit call to shutdown stream."));
+  public final void shutdown() {
+    // Don't lock on "this" before poisoning the request observer since otherwise the observer may
+    // be blocking in send().
+    requestObserver.poison();
+    synchronized (this) {
+      if (!isShutdown) {
+        isShutdown = true;
+        debugMetrics.recordShutdown();
+        shutdownInternal();
+      }
     }
   }
 
-  private void setLastError(String error) {
-    lastError.set(error);
-    lastErrorTime.set(DateTime.now());
-  }
-
-  public static class WindmillStreamShutdownException extends RuntimeException {
-    public WindmillStreamShutdownException(String message) {
-      super(message);
+  protected abstract void shutdownInternal();
+
+  /** Returns true if the stream was torn down and should not be restarted internally. */
+  private synchronized boolean maybeTearDownStream() {
+    if (requestObserver.hasReceivedPoisonPill()
```

Comment:
what is this receivedPoisonPill check guarding against?
It is racy because we poison outside of the synchronized block so we could have

T1: notices unrelated stream failure, passes this check and isn't poisoned, starts calling onNewStream
T2: calls shutdown, poisons request observer
T1: calls requestObserver.reset() gets exception due to poison.

Instead of checking the poison here, it seems like we should just handle the exception due to reset failing as that covers both cases.

Reply:
done, realized we weren't calling `break` on the `WindmillStreamShutdownException` in `startStream()`